### ISSUE TITLE
Refactor dialogs to implement MVVM pattern

### ISSUE THREAD

**Issue by Herdo** [ErikEJ/EFCorePowerTools#116]
The current dialogs in the `EFCorePowerTools` project include business logic in their background classes. This makes the dialogs not really testable, as business logic requires an instance of the dialog to be created.

Implement the MVVM pattern for the dialogs to seperate the UI logic from the business logic.

### Requires:
- New shared project required (probably .NET Standard 2.0), e.g. `EFCorePowerTools.Shared`
- Adding dependency injection
- Loose coupling between model (data access), view model (view logic, model connector) and view (dialog only)

### Benefits:
- Components (dialogs and other layers) will be testable/better testable
- Stability through more tests
- Easier extensibility through seperate layers

**Comment by Herdo** [comment]
@ErikEJ I refactored the "About" dialog, which ends up in a very similiar looking dialog:
**Current:**
![grafik](https://user-images.githubusercontent.com/651191/47607341-ca238680-da1e-11e8-9b51-c3023f4d542d.png)

**Refactored:**
![grafik](https://user-images.githubusercontent.com/651191/47607346-d4458500-da1e-11e8-9c35-4969c7ebbef7.png)

Feel free to take a look at the commit: https://github.com/Herdo/EFCorePowerTools/commit/0d7ca289ac26cb47d0a3ad1d854e292a52dc8366 Of cause, there's some ground work in there, that'll be reused for other dialogs.

Yet, the refactoring to MVVM results in smaller and testable units. 👍 
Of cause, the refactored version is fully functional.

I would go for the remaining dialogs, if you like the changes.

**Comment by ErikEJ** [comment]
This is excellent, thanks a lot - and I really like small, targeted PRs!

**Comment by ErikEJ** [comment]
I had a look, and really like the changes, please keep future PRs as small as possible (understand the size because of groundwork here)

**Comment by Herdo** [comment]
Would you be fine with one PR for each dialog?
I think it won't make much sense making too small ones, as main probably won't build then...

I can, however, try to make small commits inside each pull request. If that's fine by you.
I guess that's still easier to review.

I'd add multiple PRs after changing all dialogs to get the logic covered with tests.

**Comment by ErikEJ** [comment]
A PR per dialog would be just fine 😎👍

**Comment by Herdo** [comment]
@ErikEJ Currently refactoring the `PickServerDatabaseDialog`. Would you have any problems with me renaming it?

Proposal: `ReverseEngineerSourceSelectionDialog`.
I'm asking because of this line: `Telemetry.TrackPageView(nameof(PickServerDatabaseDialog));`

**Comment by ErikEJ** [comment]
@Herdo Please go ahead - but please do so in a separate PR (also notice that Telemetry is currently disabled due to some performance issues)

**Comment by Herdo** [comment]
Yeah, will do it in a separate PR, I also saw that there are multiple usages, not just for reverse engineer.

Could you elaborate some more on the performance issues of Telemetry? I could take a look at it after this dialog.

**Comment by ErikEJ** [comment]
This was the telemetry issue: https://github.com/ErikEJ/SqlCeToolbox/issues/575

**Comment by tomzre** [comment]
It seems the lates version from master doesn't work. When I want to any of the tools.


``` 
System.Windows.Markup.XamlParseException: 'Could not load file or assembly 'System.Windows.Interactivity, PublicKeyToken=31bf3856ad364e35' or one of its dependencies. The system cannot find the file specified.' 
```

```
FileNotFoundException: Could not load file or assembly 'System.Windows.Interactivity, PublicKeyToken=31bf3856ad364e35' or one of its dependencies. The system cannot find the file specified.
```

Do I need to add something to the project to start it working?

**Comment by ErikEJ** [comment]
Thanks for  reporting this, it seems like a bug in the refactored dialog : @Herdo

**Comment by ErikEJ** [comment]
>> but it'll break compability.

How?

**Comment by Herdo** [comment]
@ErikEJ 

### Dialog and ViewModel Creation

`IPickTablesDialog.IncludeTables()` sets the `IncludeTables` property of the `PickTablesViewModel` to `true`, by default it is `false`.

#### `ServerDgmlHandler.cs`, line 84

    var ptd = _package.GetView<IPickTablesDialog>()
                      .AddTables(predefinedTables);

=> `IncludeTables` is `false`.

#### `ReverseEngineerHandler.cs`, line 125

    var ptd = _package.GetView<IPickTablesDialog>()
                      .IncludeTables()
                      .AddTables(predefinedTables)
                      .PreselectTables(preselectedTables);

=> `IncludeTables` is `true`.

### Dialog and ViewModel Result

To fetch the result of the dialog, `IncludeTables` gets evaluated.

#### `PickTablesViewModel.cs`, line 234

    TableInformationModel[] IPickTablesViewModel.GetResult()
    {
        return (from t in Tables.Where(m => m.Model.HasPrimaryKey)
                where IncludeTables && t.IsSelected || !IncludeTables && !t.IsSelected
                select t.Model)
           .ToArray();
    }

If I read it correctly, it returns all tables that have a primary key, depending if they are selected or not selected, and the state of `IncludeTables`.

`IncludeTables == true`  => returns all tables that are selected.
`IncludeTables == false` => returns all tables that are **not** selected.

So in conclusion, the `ServerDgmlHandler` is a blacklist selection. Which is why I asked the other day what's the reason for the `IncludeTables` property, because I don't understand why there would be a negative selection

**Comment by ErikEJ** [comment]
Got it, thanks! I will have a closer look

**Comment by ErikEJ** [comment]
We need to just keep it! 😄

**Comment by Herdo** [comment]
@ErikEJ That reasoning is beyond... 🤣 
Well, a suggestion maybe:
We could remove the code from the `PickTablesViewModel` as it only makes things harder there.
Instead of this, we could/should adjust the caller (`ServerDgmlHandler`):

    var ptd = _package.GetView<IPickTablesDialog>()
                      .AddTables(predefinedTables);

    var pickTablesResult = ptd.ShowAndAwaitUserResponse(true);
    if (!pickTablesResult.ClosedByOK) return;

    var unselectedTables = predefinedTables.Except(pickTablesResult.Payload).ToArray();

This would move the logic from the ViewModel to the caller, where the actual information is required.
I think this would make it easier to read and to understand.

Of cause, the other call with `IncludeTables` would have to be adjusted.

**Comment by ErikEJ** [comment]
Sound like a good plan!

**Comment by ErikEJ** [comment]
@Herdo When debugging reverse engineering based on latest master, I now get this error - can you help? I just installed the Blend SDK component, not Blend itself...

  | Name | Value | Type
-- | -- | -- | --
  | $exception | error CS1705: Assembly 'EFCorePowerTools' with identity 'EFCorePowerTools, Version=1.0.0.0, Culture=neutral, PublicKeyToken=d9645e6262c2eca7' uses 'System.ValueTuple, Version=4.0.3.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51' which has a higher version than referenced assembly 'System.ValueTuple' with identity 'System.ValueTuple, Version=4.0.1.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51' |  
  | Name | Value | Type
-- | -- | -- | --
  | $exception | error CS1705: Assembly 'EFCorePowerTools' with identity 'EFCorePowerTools, Version=1.0.0.0, Culture=neutral, PublicKeyToken=d9645e6262c2eca7' uses 'System.ValueTuple, Version=4.0.3.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51' which has a higher version than referenced assembly 'System.ValueTuple' with identity 'System.ValueTuple, Version=4.0.1.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51' |

**Comment by ErikEJ** [comment]
@Herdo I am running VS 15.7.4 - which version do you use?

**Comment by Herdo** [comment]
I'm currently using 15.9.2.
With Blend installed.
The component not being installed.

The `System.ValueTuple` exceptions look like the issue from a few days ago: https://github.com/ErikEJ/EFCorePowerTools/pull/134

**Comment by ErikEJ** [comment]
Thanks, I will try on a different machine with a newer VS

**Comment by ErikEJ** [comment]
I guess we cannot close this one yet?

**Comment by Herdo** [comment]
There's two more dialogs left.
I could do this in this issue, or do them in separate issues. IDC.

**Comment by Herdo** [comment]
@ErikEJ Just FYI: starting with the next dialog now, after the break ✈️

**Comment by ErikEJ** [comment]
Nice! Any comments re my recent bug fix?

**Comment by Herdo** [comment]
@ErikEJ Yeah. I think I had to change the || and && several times to figure out what is right. As I still have no idea what's the best thing there, just leave it there until somebody complains ^^ I wouldn't figure out that it was this poriton of the code after reading the bug report.

**Comment by ErikEJ** [comment]
Yes, took me a while to figure it out!

**Comment by Herdo** [comment]
@ErikEJ I think we may close this for now?

**Comment by ErikEJ** [comment]
MigrationsDialog?

**Comment by Herdo** [comment]
Ups, yeah - one more to go ^^

**Comment by ErikEJ** [comment]
Finally closing this, thanks for the effort!
